### ISSUE TITLE
Vuex "dispatch" causing memory leak - related to v8 on iOS?

### ISSUE THREAD

**Issue by epaterlini** [NativeScript/NativeScript#9151]
### Version
7.0.3

### Reproduction link
[https://play.nativescript.org/?template=play-vue&id=zJW5YI&v=3](https://play.nativescript.org/?template=play-vue&id=zJW5YI&v=3)
### Platform and OS info
iOS 14.3 / NativeScript 7.0.3 / XCode 12.3

### Steps to reproduce
- Create a minimal app with NativeScript-Vue and Vuex
- In the "mounted" hook of the main component create a setInterval function that dispatches an action, eg:
setInterval(() => this.$store.dispatch('test'), 1)
where 'test' is an empty action in the store.
(I know that 1 ms is very aggressive but serves the purpose of showing the mem leak clearly)

### What is expected?
No memory leaks should happen

### What is actually happening?
Memory keeps growing indefinitely

### Comments
On Android the leak does not occur.
Not happening with tns-core-modules 6.5.0 and tns-ios 6.5.0, so the cause might be the introduction of v8 also on iOS.

**Comment by epaterlini** [comment]
Instruments screenshot WITHOUT "dispatch" call:
![image](https://user-images.githubusercontent.com/46331168/104905986-6b41e300-5983-11eb-8ded-b34da701a41c.png)

Instruments screenshot WITH "dispatch" call enabled:
![image](https://user-images.githubusercontent.com/46331168/104906052-798fff00-5983-11eb-834e-21e71661f299.png)

**Comment by schnapzz** [comment]
I've got the exact same experience in a pure JS N{7} project. My issue arrise when downloading sufficiently large files from a webserver. 

To a complete Vue novice, is the dispatch related to a web api call? Would help my own troubleshooting :-)

**Comment by epaterlini** [comment]
This is the code from Vuex store.js regarding the "dispatch" function.
Nothing special but the fact that it uses Promises....

`dispatch (_type, _payload) {
    // check object-style dispatch
    const {
      type,
      payload
    } = unifyObjectStyle(_type, _payload)

    const action = { type, payload }
    const entry = this._actions[type]
    if (!entry) {
      if (__DEV__) {
        console.error(`[vuex] unknown action type: ${type}`)
      }
      return
    }

    try {
      this._actionSubscribers
        .slice() // shallow copy to prevent iterator invalidation if subscriber synchronously calls unsubscribe
        .filter(sub => sub.before)
        .forEach(sub => sub.before(action, this.state))
    } catch (e) {
      if (__DEV__) {
        console.warn(`[vuex] error in before action subscribers: `)
        console.error(e)
      }
    }

    const result = entry.length > 1
      ? Promise.all(entry.map(handler => handler(payload)))
      : entry[0](payload)

    return new Promise((resolve, reject) => {
      result.then(res => {
        try {
          this._actionSubscribers
            .filter(sub => sub.after)
            .forEach(sub => sub.after(action, this.state))
        } catch (e) {
          if (__DEV__) {
            console.warn(`[vuex] error in after action subscribers: `)
            console.error(e)
          }
        }
        resolve(res)
      }, error => {
        try {
          this._actionSubscribers
            .filter(sub => sub.error)
            .forEach(sub => sub.error(action, this.state, error))
        } catch (e) {
          if (__DEV__) {
            console.warn(`[vuex] error in error action subscribers: `)
            console.error(e)
          }
        }
        reject(error)
      })
    })
  }
`

**Comment by epaterlini** [comment]
The problem is with the Promises' "resolve" or "reject" .
Calling those functions causes a memory leak. Not calling them is ok.

Reproduction code:
`dispatch() {
  return new Promise((resolve, reject) => {
    resolve()
  })
}
setInterval(() => {
  dispatch().then(() => {})
}, 1)`

**Comment by schnapzz** [comment]
This is reproducable for me as well using a plain JS hello world template and the latest versions of nativescript- core, ios and cli. A lot of thanks to you @epaterlini ! You've saved me so many hours of troubleshooting!

**Comment by shirakaba** [comment]
@epaterlini @schnapzz is this independent of Vuex? i.e. would we see a memory leak if one simply returned a Promise on an interval, without dispatching a Vuex action?

**Comment by epaterlini** [comment]
> @epaterlini @schnapzz is this independent of Vuex? i.e. would we see a memory leak if one simply returned Promises on an interval, without dispatching a Vuex action?
> 
> EDIT: Another thing to check: would we see a memory leak if one invoked a large number of Promises (e.g. using a for loop) without using `setInterval()`?

Yes this seems independent from Vuex, but related simply to calling Promise.resolve .
I suppose that the behaviour would be the same without calling setInterval, but the problem using a for loop would be that the code will become unresponsive and is difficult to monitor the memory usage in such conditions (IMHO).

**Comment by shirakaba** [comment]
Wasn’t suggesting an endless for loop; just say a burst or twenty or so Promises so that it would be easy to see a leak show up in the Devtools.

**Comment by epaterlini** [comment]
Hello @shirakaba,
sorry for misunderstanding.
Yes this happens also with this simple repeating loop and is much more visible:

```
var dispatch = function() {
  setTimeout(function() {
    for(var i=0; i < 1000; i++) {
      var p = new Promise((resolve, reject) => { resolve() })
    }
    dispatch()
  }, 50)
}
dispatch()
```

**Comment by brysem** [comment]
I am experiencing this issue in a production app as well. Anyone have an idea of how long a fix takes to implement or a temporary solution? I am afraid I can't help with fixing the source code in ns-v8-ios-runtime.

**Comment by shirakaba** [comment]
@brysem Follow https://github.com/NativeScript/ns-v8ios-runtime/issues/100 for progress on the investigation into this issue.
